lib/class-wp-theme-json-resolver-gutenberg.php

```diff
@@ -563,6 +584,21 @@ public static function get_user_data() {
 			}
 		}
 
+		/*
+		 * The Separator block uses different CSS properties for its color depending
+		 * on how it is being rendered e.g. as "content" for the Dots style, or
+		 * as a border etc.
+		 *
+		 * Uses are only presented with a single color control for background. Any
+		 * selection of a background color should be applied to the other paths
+		 * so it can be honored.
+		 */
+		$separator_color = $config['styles']['blocks']['core/separator']['color']['background'] ?? null;
```

Comment:
Noticed over here: https://github.com/WordPress/gutenberg/issues/57297#issuecomment-2513233950

What do you think we should do about gradients? If anything at all?

They are applied to the `background` prop, but don't really have any effect. And `linear-gradient` is invalid for `color` and `border-color` anyway as far as I see 🤔

Reply:
I'll dig into this one tomorrow and see if there's anything we can smooth out.

Reply:
I've spent some time on an archaeological dig today around the addition of gradient support to the Separator block. 

The TL;DR is that I believe it was simply mistakenly added. It wasn't referred to at all on the [original PR](https://github.com/WordPress/gutenberg/pull/38428/) and doesn't appear to have been tested in any way either. From what I can tell gradients have been broken since their adoption.

---

So...here's the long version of my current understanding 😅 

- Global Styles will generate styles targeting the `background` CSS property
- The Separator block's `hr` element will inherit `currentColor` for its `color` CSS value
- In the editor, the Separator block has a tiny amount of padding added which exposes some `background`
- On the frontend, the Separator has no padding normally so no background is visible
- Even in the editor where there is some background visible it's overlaid by the `color` property effectively masking any gradient applied here
- If the `background` property has a `linear-gradient` value at all, copying that to `border-color` or `color` results in an invalid CSS rule and falls back to `currentColor`
- The Separator block was updated in https://github.com/WordPress/gutenberg/pull/38428 which skipped color block support serialization but it didn't take gradients into account at all
- The approach in [#38428](https://github.com/WordPress/gutenberg/pull/38428) means 
     - gradient classes aren't considered and remain on the block wrapper 
     - any preset `gradient`  isn't retrieved and processed like the `backgroundColor` value
     - when [`customColor` is determined](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/separator/save.js?plain=1#L17) it only takes into account `style.color.background` forgetting about `style.color.gradient` 
- There's one default theme (TwentyTwenty) that appears to leverage `linear-gradient` while it sort of works the `//` characters don't get the gradient and hovering the block in the editor breaks the `//` onto separate lines
- A theme could style the Separator block's `hr` element to not use border or color, give it a height or padding and then apply any background style including gradients. 

At the end of all of that, I'm not sure if this is fixable to a degree where the Separator block could reliably have gradients applied to it. If it was "fixed" it might only be for use cases where the Separator block has had its default styles overridden or has a block style that makes `color` transparent, removes `border`, and gives it a visible height.

I'd like to see gradient support removed from the block. Given its current broken state and that it has been that way since its inception, I don't think there's much of a backwards compatibility argument to be made.  Removing the gradients support at least lowers complexity and maintenance burden moving forward.

What do you think @ramonjd?

Reply:
Amazing digging, thank you!

> I'd like to see gradient support removed from the block. Given its current broken state and that it has been that way since its inception, I don't think there's much of a backwards compatibility argument to be made. Removing the gradients support at least lowers complexity and maintenance burden moving forward.

+1

In any case, since it's a long-standing issue, and doesn't affect the intention of this PR it's a "tomorrow" thing. 

I appreciate the time you spent researching. It'll be useful to whoever picks it up.

Reply:
If we do go with removing support it should only be deleting a single line from the block.json. Despite that, still worth a separate PR to debate and document the decision. I'll spin that up after I work through the possible follow-up for the [issue](https://github.com/WordPress/gutenberg/pull/67269#issuecomment-2516012104) Dan flagged below.